cunumeric/array.py

```diff
@@ -2116,6 +2116,131 @@ def dumps(self):
+        if user_sizes:
+            # Zero padding if any of the user sizes is larger than input
+            zeropad_input = self
+            if np.any(np.greater(fft_s, fft_input.shape)):
+                # Create array with superset shape, fill with zeros,
+                # and copy input in
+                max_size = tuple(np.maximum(fft_s, fft_input.shape))
+                zeropad_input = ndarray(shape=max_size, dtype=fft_input.dtype)
+                zeropad_input.fill(0)
+                slices = tuple(slice(0, i) for i in fft_input.shape)
+                zeropad_input._thunk.set_item(slices, fft_input._thunk)
+
+            # Slicing according to final shape
+            fft_input_shape = list(fft_input.shape)
+            for idx, ax in enumerate(fft_axes):
+                fft_input_shape[ax] = s[idx]
+            fft_input_shape = tuple(fft_input_shape)
+            # TODO: always copying is not the best idea,
+            # sometimes a view of the original input will do
+            slices = tuple(slice(0, i) for i in fft_s)
+            fft_input = ndarray(
+                shape=fft_input_shape,
+                thunk=zeropad_input._thunk.get_item(slices),
+            ).copy()
+            fft_output_shape = fft_input_shape
+
+        # R2C/C2R require different output shapes
+        if fft_output_type != self.dtype:
+            fft_output_shape = list(fft_output_shape)
+            # R2C/C2R dimension is the last axis
+            lax = fft_axes[-1]
+            if direction == FFTDirection.FORWARD:
+                fft_output_shape[lax] = fft_output_shape[lax] // 2 + 1
+            else:
+                if user_sizes:
+                    fft_output_shape[lax] = s[-1]
+                else:
+                    fft_output_shape[lax] = 2 * (fft_input.shape[lax] - 1)
+            fft_output_shape = tuple(fft_output_shape)
+
+        # Execute FFT backend
+        out = ndarray(
+            shape=fft_output_shape,
+            dtype=fft_output_type,
+        )
+        fft_input._thunk.fft(out._thunk, fft_axes, kind, direction)
+
+        # Normalization
+        fft_norm = FFTNormalization.from_string(norm)
+        do_normalization = any(
+            [
+                fft_norm == FFTNormalization.ORTHOGONAL,
+                fft_norm == FFTNormalization.FORWARD
+                and direction == FFTDirection.FORWARD,
+                fft_norm == FFTNormalization.INVERSE
+                and direction == FFTDirection.INVERSE,
+            ]
+        )
+        if do_normalization:
+            if direction == FFTDirection.FORWARD:
+                norm_shape = fft_input.shape
+            else:
+                norm_shape = out.shape
+            norm_shape_along_axes = [norm_shape[ax] for ax in fft_axes]
+            factor = np.product(norm_shape_along_axes)
+            if fft_norm == FFTNormalization.ORTHOGONAL:
+                factor = np.sqrt(factor)
+            return out * 1.0 / factor
```

Comment:
What's the reason for the multiplication with 1.0? You may think that looks innocuous, but that will launch tasks. If what you wanted is to multiply `out` with `1.0 / factor`, you may want to put parentheses to make sure of it.